File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // Two Pointers Medium Solution
    // T: O(n + n) S: O(1)
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* currA = headA;    // Pointers at the start of the shorter list
        ListNode* currB = headB;
        while(currA != currB){
            currA = (currA == nullptr) ? headB : currA->next;
            currB = (currB == nullptr) ? headA : currB->next;
        }
        return currA;
    }
};

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) {
        // Store nodes from LLB to a hashset
        HashSet<ListNode> nodesInB = new HashSet<ListNode>();
        while(headB != null){
            nodesInB.Add(headB);
            headB = headB.next;
        }
        // Check if a node in LLB is in LLA
        while(headA != null){
            if(nodesInB.Contains(headA)){
                return headA;
            }
            headA = headA.next;
        }
        return null;
    }
}